spam/spam_tutorial.py

```diff
@@ -140,234 +140,400 @@
 # * *Third-party models*: using an pre-trained model (usually a model for a different task than the one at hand)
 # * *Distant supervision*: using external knowledge base
 # * *Crowdworker labels*: treating each crowdworker as a black-box function that assigns labels to subsets of the data
+
+# %% [markdown]
+# ### Recommended practice for LF development
+
+# %% [markdown]
+# Typical LF development cycles include multiple iterations of ideation, refining, evaluation, and debugging.
+# A typical cycle consists of the following steps
+#
+# 1. Look at examples to generate ideas for LFs
+# 1. Write an initial version of an LF
+# 1. Estimate its performance by looking at examples in the training set or dev set
+# 1. Refine and debug to improve coverage or accuracy as necessary
+#
+# Labeling function development is a creative process.
+# However, there are a few principles that we highly recommend using when developing labeling functions.
+#
+# * **Use the unlabeled training set for ideation whenever possible.** Reserve labeled examples for evaluation in order to avoid overfitting LFs or models.
+# * **Develop labeling functions in isolation.** Avoid using a "boosting" approach, trying to develop LFs specifically to interact with behavior of existing LFs. If you use the existing LFs to identify uncovered areas of the data, generalize the observed pattern rather than writing an LF to cover those exact examples.
+# * **Prioritize accuracy over coverage, up to very high accuracy.** Very accurate LFs often result in higher performance models than ones with more coverage. However, coverage is often more difficult to obtain. So choose a 90% accuracy and 30% coverage version of an LF over a 95% accuracy and 10% coverage version. But choose an 80% accuracy LF and 20% coverage version of an LF over a 65% accuracy and 40% coverage version.
```

Comment:
Should we be this prescriptive? People are going to really grab onto whatever we say, and I'm not sure you really should always take the 90/30 over the 95/10; depends on how high you need performance to go, how many other LFs you have, etc. Maybe we can express it in terms of the more general rule that the discriminative model will improve coverage more than precision? idk though, since a lot of people will probably find these numbers very helpful as guideposts.

Reply:
i'll add a general "rules are meant to be broken" thing, and point to these